crates/sui-source-validation/src/lib.rs

```diff
@@ -41,119 +44,321 @@ pub struct BytecodeSourceVerifier<'a> {
-    /// the bytecode at the address specified on the Sui network we are publishing to.
-    pub async fn verify_package_root_and_deps(
-        &self,
-        compiled_package: &CompiledPackage,
-        root_on_chain_address: AccountAddress,
-    ) -> Result<(), AggregateError> {
-        self.verify_package(
-            compiled_package,
-            /* verify_deps */ true,
-            SourceMode::VerifyAt(root_on_chain_address),
-        )
-        .await
+    /// Only verify that the root package matches its on-chain version (requires that the root
+    /// package is published with its address available in the manifest).
+    pub fn root() -> Self {
+        Self::Root {
+            deps: false,
+            at: None,
+        }
     }
 
-    /// Helper wrapper to verify that all local Move package root bytecode matches
-    /// the bytecode at the address specified on the Sui network we are publishing to.
-    pub async fn verify_package_root(
-        &self,
-        compiled_package: &CompiledPackage,
-        root_on_chain_address: AccountAddress,
-    ) -> Result<(), AggregateError> {
-        self.verify_package(
-            compiled_package,
-            /* verify_deps */ false,
-            SourceMode::VerifyAt(root_on_chain_address),
-        )
-        .await
+    /// Only verify that the root package matches its on-chain version, but override the location
+    /// to look for the root package to `address`.
+    pub fn root_at(address: AccountAddress) -> Self {
+        Self::Root {
+            deps: false,
+            at: Some(address),
+        }
+    }
+
+    /// Verify both the root package and its dependencies (requires that the root package is
+    /// published with its address available in the manifest).
+    pub fn root_and_deps() -> Self {
+        Self::Root {
+            deps: true,
+            at: None,
+        }
+    }
+
+    /// Verify both the root package and its dependencies, but override the location to look for
+    /// the root package to `address`.
+    pub fn root_and_deps_at(address: AccountAddress) -> Self {
+        Self::Root {
+            deps: true,
+            at: Some(address),
+        }
     }
 
-    /// Helper wrapper to verify that all local Move package dependencies' matches
-    /// the bytecode at the address specified on the Sui network we are publishing to.
-    pub async fn verify_package_deps(
+    /// Should we verify dependencies?
+    fn verify_deps(&self) -> bool {
+        matches!(self, Self::Deps | Self::Root { deps: true, .. })
+    }
+
+    /// If the root package needs to be verified, what address should it be fetched from?
+    fn root_address(&self, package: &CompiledPackage) -> Result<Option<AccountAddress>, Error> {
+        match self {
+            Self::Root { at: Some(addr), .. } => Ok(Some(*addr)),
+            Self::Root { at: None, .. } => Ok(Some(*package.published_at.clone()?)),
```

Comment:
Noob question, this is dereferenced and then cloned? Does it need to be cloned?

Reply:
The `Result` is being cloned, then checked with `?` and in the `Ok` case, we are dereferencing the value (which in this case is actually doing a conversion from `ObjectID` to `AccountAddress`, which is a bit of an abuse of `Deref`...).

`clone` is necessary here because otherwise `?` would consume the result, which would be a partial move out of `package`.